### PR TITLE
Merge 2019-07-23-Intro to development

### PR DESCRIPTION
**Pull request by BohdanDudkaTD**
In 2019-07-23-Intro branch was created an vehicle classes hierarchy which contains both more abstract entities and concrete implementations. All classes are documented.
Also it contains tests for classes functionality.

### CODE REVIEW

**Comment by dkalinin** [comment]
@BohdanDudkaTD don't forget about Pull request description.
Also let's have a hierarchy of folders according to inheritance.

File: 2019-07-23-Intro/tests/index.html [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width">
+    <title>QUnit tests</title>
+    <link rel="stylesheet" href="https://code.jquery.com/qunit/qunit-2.9.2.css">
+</head>
+<body>
+<div id="qunit"></div>
+<div id="qunit-fixture"></div>
+<script src="https://code.jquery.com/qunit/qunit-2.9.2.js"></script>
+<script type="module" src="tests.js"></script>
```
Comment (dkalinin) [inline 2019-07-23-Intro/tests/index.html]: image
Let's create `app` folder where all classes should be moved to.
Also, let's have a folder hierarchy in `app` which represents your class hierarchy.

Comment (BohdanDudkaTD) [reply 2019-07-23-Intro/tests/index.html]: All classes are moved to the `app` folder, now folders with classes are stored according to the inheritance hierarchy.
`tests` and `app` folders are in the root of `2019-07-23-Intro` folder.

File: 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * Check if this helicopter can carry another vehicle (or extended object) by comparison this object`s carryingVolume
+     * and carryingWeight with volume and weight of passed as parameter vehicle.
+     *
+     * @public
+     * @param {Vehicle} vehicle checked if it can be carried by instance of this class.
+     * @returns {boolean} true if passed in vehicle object can be carried by this object, otherwise - false.
+     */
+    canCarry(vehicle) {
+        return super.canCarry(vehicle);
```
Comment (dkalinin) [inline 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js]: This overriding is quite dummy. We don't need it.

File: 2019-07-23-Intro/app/vehicle/aircraft/index.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     * @param {string} name - string value of aircraft`s name.
+     * @param {number} carryingVolume - integer or floating number which describes maximum volume in cubic meters  which
+     * can by carried by this aircraft.
+     * @param {number} carryingWeight - integer or floating number which describes maximum weight in tonnes  which
+     * can by carried by this aircraft.
+     * @param {number} weight - integer or floating number which describes weight of this aircraft in tonnes.
+     * @param {number} volume - integer or floating number which describes volume of this aircraft in cubic meters.
+     * @param {string} movingEnvironment - environment in which this aircraft can move.
+     */
+    constructor(name, carryingVolume, carryingWeight, weight, volume, movingEnvironment) {
```
Comment (dkalinin) [inline 2019-07-23-Intro/app/vehicle/aircraft/index.js]: Why do we pass movingEnvironment to Aircraft constructor? Isn't it known?

File: 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+     *
+     * @public
+     * @param {string} name - string value of helicopter`s name.
+     * @param {number} carryingVolume - integer or floating number which describes maximum volume in cubic meters  which
+     * can by carried by this helicopter.
+     * @param {number} carryingWeight - integer or floating number which describes maximum weight in tonnes  which
+     * can by carried by this helicopter.
+     * @param {number} weight - integer or floating number which describes weight of this helicopter in tonnes.
+     * @param {number} volume integer or floating number which describes volume of this helicopter in cubic meters.
+     * @param {string} movingEnvironment - environment in which this helicopter can move.
```
Comment (dkalinin) [inline 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js]: Let's have a separate class for 'movingEnvironment'.

Comment (BohdanDudkaTD) [reply 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js]: Should it be some kind of enum?

File: 2019-07-23-Intro/app/vehicle/aircraft/index.js (hunk 3 above)
Comment (BohdanDudkaTD) [reply 2019-07-23-Intro/app/vehicle/aircraft/index.js]: Do you mean it should be a const field?

File: 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js (hunk 2 above)
Comment (BohdanDudkaTD) [reply 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js]: Removed unnecessary overriding.

**Comment by BohdanDudkaTD** [comment]
@dkalinin, fixed, PTAL.

Fixed issues:
- extract possible moving environments for vehicles to separate class (MovingEnvironmentEnum).
Add relevant tests;
- also movingEnvironment is now not passed in constructor but assigned at field creation;
- remove unnecessary overriding;
- add unique fields for few classes;

**Comment by BohdanDudkaTD** [comment]
@dkalinin fixed, PTAL.

Fixed issues:
- update JS docs, add override anotation;
- remove unnecessary fields from classes, move movementEnvironment field initialization to constructor;
- now Vehicle class can't be instantiated, get movingEnvironment can't be called directly;
- rename MovingEnvironmentEnum, remove unnecessary indefinite field;
- refactor code style, test messages;

**Comment by BohdanDudkaTD** [comment]
Updates:

- add classes of predefined vehicles;
- add parameterized tests.

File: 2019-07-23-Intro/app/vehicle/bicycle/index.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     * Get environment in which current vehicle can move. In this case it`s ground environment.
+     *
+     * @public
+     * @override
+     * @returns {MovingEnvironment} enum`s {@link MovingEnvironment.GROUND} constant with string field.
+     */
+    get movingEnvironment() {
+        return this._movingEnvironment;
```
Comment (dkalinin) [inline 2019-07-23-Intro/app/vehicle/bicycle/index.js]: There's no need to have a private property `_movingEnvironment ` just return `MovingEnvironment.GROUND`.

File: 2019-07-23-Intro/tests/fishBoatTests.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import {BaylinerElementF16} from "../app/vehicle/ship/fishboat/baylinerelementf16";
+import {MovingEnvironment} from "../app/vehicle/movingenvironment";
+
+QUnit.module("Fish boat", {
+    before: function () {
+        this.testObject = [
+            {
+                vehicleInstance: new CannondaleCatalyst2("Speed"),
+                expectedCanCarryResult: true,
+                assertionMessage: "Fish boat should carry bicycle."
```
Comment (dkalinin) [inline 2019-07-23-Intro/tests/fishBoatTests.js]: Cannondale is a bicycle. Check other assertionMessages too.

File: 2019-07-23-Intro/tests/fishBoatTests.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            {
+                vehicleInstance: new PorscheBoxsterSpyder("FastGuy"),
+                expectedCanCarryResult: false,
+                assertionMessage: "Fish boat should not carry roadster."
+            }];
+    }
+});
+QUnit.test("Test fish boat`s canCarry.", function (assert) {
+    const fishBoat = new BaylinerElementF16("Fishing");
+    this.testObject.forEach(element => {
```
Comment (dkalinin) [inline 2019-07-23-Intro/tests/fishBoatTests.js]: It's better to create create `QUnit.test`in function which iterates testObject.

File: 2019-07-23-Intro/tests/fishBoatTests.js (hunk 6 above)
Comment (BohdanDudkaTD) [reply 2019-07-23-Intro/tests/fishBoatTests.js]: Could you, please, explain more detailed where exactly is the mistake?
Here a fish boat is the testable object and a bicycle is one of vehicle`s instances which should be carried by a fish boat.

**Pull request closed without merging**